group-coordinator/src/main/java/org/apache/kafka/coordinator/group/ConsumerGroupMetadataManager.java

```diff
@@ -0,0 +1,1356 @@
+import static org.apache.kafka.coordinator.group.metrics.GroupCoordinatorMetrics.CONSUMER_GROUP_REBALANCES_SENSOR_NAME;
+import static org.apache.kafka.coordinator.group.modern.ModernGroupMember.hasAssignedPartitionsChanged;
+
+public class ConsumerGroupMetadataManager {
+
+    private final GroupStore groupStore;
+    private final GroupMetaManagerHelper helper;
+
+    /**
+     * The default consumer group assignor used.
+     */
+    private final ConsumerGroupPartitionAssignor defaultConsumerGroupAssignor;
+
+    /**
+     * The supported consumer group partition assignors keyed by their name.
+     */
+    private final Map<String, ConsumerGroupPartitionAssignor> consumerGroupAssignors;
+
+    /**
+     * The metadata refresh interval.
+     */
+    private final int consumerGroupMetadataRefreshIntervalMs;
+
+    /**
+     * The maximum number of members allowed in a single consumer group.
+     */
+    private final int consumerGroupMaxSize;
+
+    /**
+     * The default heartbeat interval for consumer groups.
+     */
+    private final int consumerGroupHeartbeatIntervalMs;
+
+    /**
+     * Initial rebalance delay for members joining a classic group.
+     */
+    private final int classicGroupInitialRebalanceDelayMs;
+
+    /**
+     * The maximum number of members allowed in a single classic group.
+     */
+    private final int classicGroupMaxSize;
+
+    /**
+     * The config indicating whether group protocol upgrade/downgrade is allowed.
+     */
+    private final ConsumerGroupMigrationPolicy consumerGroupMigrationPolicy;
+
+    /**
+     * The group manager.
+     */
+    private final GroupConfigManager groupConfigManager;
+
+    /**
+     * The default session timeout for consumer groups.
+     */
+    private final int consumerGroupSessionTimeoutMs;
+
+    /**
+     * The coordinator metrics.
+     */
+    private final GroupCoordinatorMetricsShard metrics;
+
+    /**
+     * The system time.
+     */
+    private final Time time;
+
+    /**
+     * The system timer.
+     */
+    private final CoordinatorTimer<Void, CoordinatorRecord> timer;
+
+    /**
+     * The log context.
+     */
+    private final LogContext logContext;
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+
+    public ConsumerGroupMetadataManager(
+        GroupStore groupStore,
+        List<ConsumerGroupPartitionAssignor> consumerGroupAssignors,
+        int consumerGroupMetadataRefreshIntervalMs,
+        int classicGroupInitialRebalanceDelayMs,
+        int consumerGroupMaxSize,
+        int consumerGroupHeartbeatIntervalMs,
+        int consumerGroupSessionTimeoutMs,
+        int classicGroupMaxSize,
+        GroupConfigManager groupConfigManager,
+        ConsumerGroupMigrationPolicy consumerGroupMigrationPolicy,
+        LogContext logContext,
+        GroupCoordinatorMetricsShard metrics,
+        Time time,
+        CoordinatorTimer<Void, CoordinatorRecord> timer) {
+        this.groupStore = groupStore;
```

Comment:
nit: Probably more legible with a line break before the `) {` so the method body is distinct from the argument declarations.